ui/src/components/CreateClusterForm.js

```diff
@@ -0,0 +1,288 @@
+import { withStyles } from 'material-ui/styles';
+import { InputLabel } from 'material-ui/Input';
+import TextField from 'material-ui/TextField';
+import Typography from 'material-ui/Typography';
+import { MenuItem } from 'material-ui/Menu';
+import { FormControl } from 'material-ui/Form';
+import Select from 'material-ui/Select';
+import Button from 'material-ui/Button';
+import { CircularProgress } from 'material-ui/Progress';
+
+import { createApiUrl } from '../net'
+
+
+const DEFAULT_MASTER_MACHINE_TYPE = "n1-standard-4";
+const DEFAULT_WORKER_MACHINE_TYPE = "n1-standard-2";
+const DEFAULT_WORKER_DISK_SIZE = 200;
+
+// Regex for GCP Project names.
+const projectRegex = /^[a-z][a-z0-9_-]{0,50}[a-z0-9]$/;
+
+// Regex for cluster names.
+const clusterRegex = /^[a-z][a-z0-9-]{0,25}[a-z0-9]$/;
+
+const validMachineTypes = [
+  "n1-standard-2",
+  "n1-standard-4",
+  "n1-standard-8"];
+
+const validDiskSizes = [200, 500, 1000];
+
+/**
+ * Spinner that only shows when there's a request in progress.
+ * @props requestInProgress bool indicates request status.
+ */
+class HiddenSpinner extends React.Component {
+  render() {
+    if (this.props.requestInProgress) {
+      return (
+        <span>
+          <CircularProgress color="secondary" />
+        </span>);
+    } else {
+      return <span/>
+    }
+  }
+}
+
+
+/**
+ * Render a form used to configure and launch a new cluster.
+ * @props closeFormOnSuccess function callback to modal to close the
+ *        form and refresh clusters from the API.
+ * @props closeFormOnFailure function callback to modal to close the
+ *        form and display the error.
+ * @props googleAuthToken string auth token from oauth login.
+ */
+class UnstyledCreateClusterForm extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      clusterName: "",
+      googleProject: "",
+      machineType: DEFAULT_MASTER_MACHINE_TYPE,
+      workerMachineType: DEFAULT_WORKER_MACHINE_TYPE,
+      diskSize: 200,
+      createRequestValid: false,
+      requestInProgress: false,
+      numberOfWorkers: 0,
+    };
+  }
+
+  postStateChangeValidationCallback = () => {
+    if (projectRegex.test(this.state.googleProject)
+        && clusterRegex.test(this.state.clusterName)
+        && validMachineTypes.includes(this.state.machineType)
+        && validDiskSizes.includes(this.state.diskSize)) {
+      this.setState({createRequestValid: true});
+    } else {
+      this.setState({createRequestValid: false});
+    }
+  }
+
+  createClusterRequest = (event) => {
+    // Sets state to disable submission & and show loading spinner.
+    this.setState({
+      createRequestValid: false,
+      requestInProgress: true,
+    });
+
+    var createRequest = {
+      labels: {},
+      machineConfig: {
+        // Worker config.
+        numberOfWorkers: this.state.numberOfWorkers,
+        workerMachineType: this.state.workerMachineType,
+        numberOfPreemptibleWorkers: 0,
```

Comment:
Why disallow preemptibles? In my experience people like to use them as they are _much_ cheaper.

Reply:
I agree with the sentiment, see my comment on this PR.